internal/vpmem/mapped.go

```diff
@@ -0,0 +1,282 @@
+func (vpd *virtualPMemSurface) mapDevice(ctx context.Context, device mappedVPMem) (err error) {
+	if dev, ok := vpd.mappings[device.HostPath()]; ok {
+		dev.AddRef()
+		return nil
+	}
+
+	// Make sure that the requested offset is matching the first available one
+	memCls := memory.GetMemoryClassType(device.DeviceSize())
+	offset, err := vpd.FindFreeOffset(memCls)
+	if err != nil {
+		return err
+	}
+	if offset != device.DeviceOffset() {
+		return errors.Errorf("device offset mismatch: available=%d, requested=%d", offset, device.DeviceOffset())
+	}
+
+	// Allocate memory slot and make sure that the expected allocation happened
+	region, err := vpd.Allocate(memCls)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			vpd.Release(memCls, region.Offset())
+		}
+	}()
+
+	if region.Offset() != device.DeviceOffset() {
+		return errors.Errorf("unexpected region allocation: expected=%d, actual=%d", offset, region.Offset())
+	}
+
+	log.G(ctx).WithFields(logrus.Fields{
+		"hostPath":     device.HostPath(),
+		"mountPath":    device.MountPath(),
+		"deviceNumber": device.DeviceNumber(),
+		"deviceOffset": device.DeviceOffset(),
+		"deviceSize":   device.DeviceSize(),
+	}).Debug("mapped new device")
+
+	vpd.mappings[device.HostPath()] = device
+	return nil
+}
+
+// unmapDevice reduces reference of a specified device and takes care of releasing the allocated
+// memory when appropriate
+func (vpd *virtualPMemSurface) unmapDevice(ctx context.Context, hostPath string) error {
+	dev, ok := vpd.mappings[hostPath]
+	if !ok {
+		return ErrNotAttached
+	}
+
+	if dev.RefCount() == 1 {
+		log.G(ctx).WithFields(logrus.Fields{
+			"hostPath": hostPath,
+		}).Debug("releasing resources")
+		memCls := memory.GetMemoryClassType(dev.DeviceSize())
+		err := vpd.Release(memCls, dev.DeviceOffset())
+		if err != nil {
+			return err
+		}
+		delete(vpd.mappings, hostPath)
+		return nil
+	}
+	dev.RemoveRef()
+	return nil
+}
+
+func (mc *mappedVPMemController) FindDevice(ctx context.Context, hostPath string) (genericVPMem, error) {
+	for i := uint32(0); i < mc.vpmemMaxCount; i++ {
+		vi := mc.vpmemDevices[i]
+		if vi != nil {
+			if dev, ok := vi.mappings[hostPath]; ok {
+				log.G(ctx).WithFields(logrus.Fields{
+					"deviceNumber": i,
+					"hostPath":     dev.HostPath(),
+					"uvmPath":      dev.MountPath(),
+					"refCount":     dev.RefCount(),
+					"deviceSize":   dev.DeviceSize(),
+					"deviceOffset": dev.DeviceOffset(),
+				}).Debug("found mapped device")
+				return dev, nil
+			}
+		}
+	}
+	return nil, ErrNotAttached
+}
+
+func (mc *mappedVPMemController) FindNextDevice(ctx context.Context, hostPath string) (genericVPMem, error) {
+	fi, err := os.Stat(hostPath)
+	if err != nil {
+		return nil, err
+	}
+	devSize := pageAlign(uint64(fi.Size()))
+	if devSize > mc.vpmemMaxSizeBytes {
+		return nil, ErrMaxVPMEMLayerSize
+	}
+
+	entry := log.G(ctx).WithFields(logrus.Fields{
```

Comment:
I don't really understand this

Reply:
I was trying to avoid code duplication since I then log 2 different things